Fix ios and android builds

kp-mariappan-ramasamy opened the pull request:
When adding support for async send/recv apis in https://github.com/tun2proxy/rust-tun/pull/87 , ios and android builds got broken with following error

```
error[E0433]: failed to resolve: use of undeclared crate or module `io`
```

**Root cause:**

Both ios and android platform code did not import `std::io::self` which caused it to fail. 

This PR attempts to fix those issues.

Also I have added CI jobs to build Android and iOS targets, to find these issues earlier

Example of test fail without the fix:
https://github.com/kp-mariappan-ramasamy/rust-tun/actions/runs/10058442720/job/27801536723#step:4:95
https://github.com/kp-mariappan-ramasamy/rust-tun/actions/runs/10058442720/job/27801536214#step:4:110

kp-mariappan-ramasamy committed:
fix android and ios build

When adding support for async send/recv apis in #87, ios and android builds got broken with following error

    error[E0433]: failed to resolve: use of undeclared crate or module `io`

Root cause:
Both ios and android platform code did not import std::io::self which caused it to fail.

Updated it to use full path to fix it

kp-mariappan-ramasamy committed:
ci: add test for android

Used cargo ndk which makes compiling for android code smoother.
https://github.com/bbqsrc/cargo-ndk

kp-mariappan-ramasamy committed:
ci: add test for iOS

Used cargo-lipo, which helps in compiling for iOS target.
https://github.com/TimNN/cargo-lipo

ssrlive committed:
Update rust.yml

ssrlive committed:
ci: enable clippy for android and ios

ssrlive commented:
Seams good. but please remove the warnings.
```
 warning: unused imports: `ipaddr_to_sockaddr`, `sockaddr_union`
  --> src/platform/posix/mod.rs:18:27
   |
18 | pub(crate) use sockaddr::{ipaddr_to_sockaddr, sockaddr_union};
   |                           ^^^^^^^^^^^^^^^^^^  ^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

warning: function `ipaddr_to_sockaddr` is never used
  --> src/platform/posix/sockaddr.rs:63:15
   |
63 | pub unsafe fn ipaddr_to_sockaddr<T>(
   |               ^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(dead_code)]` on by default
```

kp-mariappan-ramasamy committed:
Fix cargo clippy warnings for android for ios

Errors:

 warning: unused imports: `ipaddr_to_sockaddr`, `sockaddr_union`
  --> src/platform/posix/mod.rs:18:27
   |
18 | pub(crate) use sockaddr::{ipaddr_to_sockaddr, sockaddr_union};
   |                           ^^^^^^^^^^^^^^^^^^  ^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

warning: function `ipaddr_to_sockaddr` is never used
  --> src/platform/posix/sockaddr.rs:63:15
   |
63 | pub unsafe fn ipaddr_to_sockaddr<T>(
   |               ^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(dead_code)]` on by default

error: this expression creates a reference which is immediately dereferenced by the compiler
  --> src/platform/android/mod.rs:29:17
   |
29 |     Device::new(&configuration)
   |                 ^^^^^^^^^^^^^^ help: change this to: `configuration`

error: this expression creates a reference which is immediately dereferenced by the compiler
  --> src/platform/ios/mod.rs:58:17
   |
58 |     Device::new(&configuration)
   |                 ^^^^^^^^^^^^^^ help: change this to: `configuration`

kp-mariappan-ramasamy commented on .github/workflows/rust.yml:
```diff
@@ -34,6 +34,8 @@ jobs:
       run: echo "Some of jobs failed" && false
 
   build_n_test_android:
+    strategy:
```
This field is applicable only for matrix builds.
Not needed for single build job.
Ref: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstrategyfail-fast

kp-mariappan-ramasamy commented:
Thanks @ssrlive I have fixed the warnings with android and ios builds. PTAL

Also fixed android and iOS CI job to run clippy properly
I think, this commit https://github.com/tun2proxy/rust-tun/pull/90/commits/4c7b5c0f6a9c728eacd5acb65132684cd694f6b0 may not be needed. But it does not hurt to include those too.

Also please let me know if you want me to rebase commits.

ssrlive commented:
OK. Please rebase it.

kp-mariappan-ramasamy commented:
@ssrlive Rebased commits for clean history. PTAL

ssrlive committed:
Bump version 2.0.4

ssrlive commented:
Thanks
